source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,587 @@
+                        direction[dim-1] = 1.;
+                      }
+
+                    // Compute the total displacement in the gravity direction,
+                    // i.e. the initial topography + any additional mesh displacement.
+                    const double displacement = direction * (displacement_values[point] + initial_topography_values[point]);
+
+                    // To project onto the tangent space of the surface,
+                    // we define the projection P:= I- n x n,
+                    // with I the unit tensor and n the unit normal to the surface.
+                    // The surface gradient then is P times the usual gradient of the shape functions.
+                    const Tensor<2, dim, double> projection = unit_symmetric_tensor<dim>() -
+                                                              outer_product(fs_fe_face_values.normal_vector(point), fs_fe_face_values.normal_vector(point));
+
+                    // Loop over the shape functions
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        // Make sure we only assemble for the y-component
+                        // TODO this is only correct for box geometries
+                        // Check that a box geometry is used
+                        AssertThrow(Plugins::plugin_type_matches<GeometryModel::Box<dim> >(this->get_geometry_model()) ||
+                                    Plugins::plugin_type_matches<GeometryModel::TwoMergedBoxes<dim> >(this->get_geometry_model()),
+                                    ExcMessage("The surface diffusion mesh deformation plugin only works for Box geometries."));
+                        if (mesh_deformation_dof_handler.get_fe().system_to_component_index(i).first == dim-1)
+                          {
+                            // Assemble the RHS
+                            // RHS = M*H_old
+                            cell_vector(i) += fs_fe_face_values.shape_value (i, point) * displacement * fs_fe_face_values.JxW(point);
+
+
+                            for (unsigned int j=0; j<dofs_per_cell; ++j)
+                              {
+                                // Make sure we only assemble for the y-component
+                                // TODO this is only correct for box geometries
+                                if (mesh_deformation_dof_handler.get_fe().system_to_component_index(j).first == dim-1)
+                                  {
+                                    // Assemble the matrix, for backward first order time discretization:
+                                    // Matrix := (M+dt*K) = (M+dt*B^T*kappa*B)
+                                    cell_matrix(i,j) +=
+                                      (
+                                        fs_fe_face_values.shape_value (i, point) * fs_fe_face_values.shape_value (j, point) +
```

Comment:
Are you concerned about the speed of this function, i.e. did you see this assembly taking a lot of time? If not, do not bother, but if yes, there is a lot of optimization potential in this assembly. For example you can precompute `phi[i] = fs_fe_face_values.shape_value (i, point)` for all i at the beginning of the loop over `point`, and then save a lot of calls to `shape_value` in line 333 and here (for i and j). Even more helpful will be precomputing `proj_grad_phi[i] = projection * fs_fe_face_values.shape_grad(i, point)` for all i, because it reduces the calls in the inner loop to the gradient calculation and the tensor product of projection and the gradient.

Reply:
I'm not sure if I did entirely what you intended, but I precompute `phi[i]` and `projected_grad_phi[i]` now.